chia/full_node/full_node_api.py

```diff
@@ -119,9 +122,56 @@ async def new_transaction(self, transaction: full_node_protocol.NewTransaction)
-            request_tx = full_node_protocol.RequestTransaction(transaction.transaction_id)
-            msg = make_msg(ProtocolMessageTypes.request_transaction, request_tx)
-            return msg
+            # If there's current pending request just add this peer to the set of peers that have this tx
+            if transaction.transaction_id in self.full_node.pending_tx_request:
+                if transaction.transaction_id in self.full_node.peers_with_tx:
+                    current_set = self.full_node.peers_with_tx[transaction.transaction_id]
+                    if peer.peer_node_id in current_set:
+                        return None
+                    current_set.add(peer.peer_node_id)
+                    return None
+                else:
+                    new_set = set()
+                    new_set.add(peer.peer_node_id)
+                    self.full_node.peers_with_tx[transaction.transaction_id] = new_set
+                    return None
+
+            self.full_node.pending_tx_request[transaction.transaction_id] = peer.peer_node_id
+            new_set = set()
+            new_set.add(peer.peer_node_id)
+            self.full_node.peers_with_tx[transaction.transaction_id] = new_set
+
+            async def tx_request_and_timeout(full_node: FullNode, transaction_id):
+                try:
+                    while True:
+                        if transaction_id not in full_node.peers_with_tx:
+                            break
+                        peers_with_tx: Set = full_node.peers_with_tx[transaction_id]
+                        if len(peers_with_tx) == 0:
+                            break
+                        peer_id = peers_with_tx.pop()
+                        assert full_node.server is not None
+                        if peer_id not in full_node.server.all_connections:
+                            continue
+                        peer = full_node.server.all_connections[peer_id]
+                        request_tx = full_node_protocol.RequestTransaction(transaction.transaction_id)
+                        msg = make_msg(ProtocolMessageTypes.request_transaction, request_tx)
+                        await peer.send_message(msg)
+                        await asyncio.sleep(20)
+                        if full_node.mempool_manager.seen(transaction_id):
+                            break
+                except asyncio.CancelledError:
+                    return
+                finally:
+                    # Always Cleanup
+                    if transaction_id in full_node.peers_with_tx:
+                        full_node.peers_with_tx.pop(transaction_id)
+                    if transaction_id in full_node.pending_tx_request:
+                        full_node.pending_tx_request.pop(transaction_id)
+
+            fetch_task = asyncio.create_task(tx_request_and_timeout(self.full_node, transaction.transaction_id))
+            self.full_node.tx_fetch_tasks.add(fetch_task)
```

Comment:
I don't see anywhere where you remove items from this list. it seems there's a leak

Reply:
good catch